**DAY1**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
!unzip -l /content/drive/MyDrive/cod_hybrid_project/cod_hybrid.zip | head -50

unzip:  cannot find or open /content/drive/MyDrive/cod_hybrid_project/cod_hybrid.zip, /content/drive/MyDrive/cod_hybrid_project/cod_hybrid.zip.zip or /content/drive/MyDrive/cod_hybrid_project/cod_hybrid.zip.ZIP.


In [ ]:
!find /content/drive/MyDrive -name "*.zip"
# Why: we're not sure of the exact path where the zip was uploaded
# What: searches your entire Drive for any .zip file
#       and prints its full path so we can use it in the next cell

/content/drive/MyDrive/SafeSite/raw.zip
/content/drive/MyDrive/cod_hybrid_project/cod_hybrid.zip


In [ ]:
!unzip /content/drive/MyDrive/cod_hybrid_project/cod_hybrid.zip \
       -d /content/drive/MyDrive/cod_hybrid_project/
# Why: Colab can't run zipped files — needs extracted folders
# What: extracts cod_hybrid.zip into cod_hybrid_project/
#       creating cod_hybrid_project/cod_hybrid/ with all code and datasets

Streaming output truncated to the last 5000 lines.
  inflating: /content/drive/MyDrive/cod_hybrid_project/cod_hybrid/data/datasets/COD10K/Train/Image/COD10K-CAM-2-Terrestrial-26-Chameleon-1729.jpg  
  inflating: /content/drive/MyDrive/cod_hybrid_project/cod_hybrid/data/datasets/COD10K/Train/Image/COD10K-CAM-2-Terrestrial-26-Chameleon-1731.jpg  
  inflating: /content/drive/MyDrive/cod_hybrid_project/cod_hybrid/data/datasets/COD10K/Train/Image/COD10K-CAM-2-Terrestrial-26-Chameleon-1732.jpg  
  inflating: /content/drive/MyDrive/cod_hybrid_project/cod_hybrid/data/datasets/COD10K/Train/Image/COD10K-CAM-2-Terrestrial-26-Chameleon-1733.jpg  
  inflating: /content/drive/MyDrive/cod_hybrid_project/cod_hybrid/data/datasets/COD10K/Train/Image/COD10K-CAM-2-Terrestrial-26-Chameleon-1734.jpg  
  inflating: /content/drive/MyDrive/cod_hybrid_project/cod_hybrid/data/datasets/COD10K/Train/Image/COD10K-CAM-2-Terrestrial-26-Chameleon-1735.jpg  
  inflating: /content/drive/MyDrive/cod_hybrid_project/cod_hy

In [ ]:
!grep "DATA_ROOT\|CKPT_DIR\|LOG_DIR\|ROOT_DIR" \
      /content/drive/MyDrive/cod_hybrid_project/cod_hybrid/configs/cod_hybrid.py
# Why: verify all 4 critical paths are correct before training
# What: searches the config file and prints only the lines
#       containing DATA_ROOT, CKPT_DIR, LOG_DIR, ROOT_DIR
#       so we can confirm they all point to the right Colab paths

# ROOT_DIR   = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
# DATA_ROOT  = os.path.join(ROOT_DIR, 'data', 'datasets')   # CAMO / COD10K here
# CKPT_DIR   = os.path.join(ROOT_DIR, 'checkpoints') 
# LOG_DIR    = os.path.join(ROOT_DIR, 'logs')
ROOT_DIR  = '/content/drive/MyDrive/cod_hybrid_project/cod_hybrid'
DATA_ROOT = '/content/drive/MyDrive/cod_hybrid_project/cod_hybrid/data/datasets'
CKPT_DIR  = '/content/drive/MyDrive/cod_hybrid_project/cod_hybrid/checkpoints'
LOG_DIR   = '/content/drive/MyDrive/cod_hybrid_project/cod_hybrid/logs'
CAMO_ROOT    = os.path.join(DATA_ROOT, 'CAMO')
COD10K_ROOT  = os.path.join(DATA_ROOT, 'COD10K')


In [ ]:
!pip install timm -q
# Why: Colab doesn't have timm pre-installed
# What: installs timm library which provides PVTv2-B2 backbone weights
# -q suppresses verbose output

In [ ]:
import os
os.chdir('/content/drive/MyDrive/cod_hybrid_project/cod_hybrid')
# Why: imports only work when running from the project root
# What: sets working directory to cod_hybrid/

!ls
# Why: confirm all folders are present before training
# What: should show configs/ data/ models/ losses/ engine/ utils/ train.py eval.py

checkpoints  data    eval.py  losses  train.py
configs      engine  logs     models  utils


In [ ]:
!ls /content/drive/MyDrive/cod_hybrid_project/
# What: shows everything inside cod_hybrid_project folder

cod_hybrid  cod_hybrid.zip


In [ ]:
!ls /content/drive/MyDrive/cod_hybrid_project/cod_hybrid/
# What: shows everything inside the cod_hybrid folder

checkpoints  configs  data


In [ ]:
!pwd
# What: shows current directory path

/content/drive/MyDrive/cod_hybrid_project/cod_hybrid


In [ ]:
!python train.py #12 hours!!
# Why: entry point for the entire pipeline
# What: builds dataloader, model, trainer and starts training saves checkpoints every 10 epochs to checkpoints/

In [ ]:
# Why: override epochs to 30 so training finishes in ~4 hours
# What: edits the config file to set NUM_EPOCHS to 30
#       so we get a usable checkpoint without running all 100 epochs now

config_path = '/content/drive/MyDrive/cod_hybrid_project/cod_hybrid/configs/cod_hybrid.py'

with open(config_path, 'r') as f:
    content = f.read()

content = content.replace(
    'NUM_EPOCHS     = 100',
    'NUM_EPOCHS     = 30'
)

with open(config_path, 'w') as f:
    f.write(content)

print('NUM_EPOCHS set to 30')

NUM_EPOCHS set to 30


In [ ]:
import torch
print(torch.cuda.is_available())
# Why: verify GPU is actually assigned before starting training
# What: prints True if GPU is available, False if not

print(torch.cuda.get_device_name(0) if torch.cuda.is_available() else "No GPU")
# Why: confirm which GPU we got
# What: prints the GPU model name e.g. "Tesla T4"

True
Tesla T4


In [ ]:
!python train.py
# Why: starts training for 30 epochs (~4 hours) (Actually it's 12 hours)
# What: saves checkpoints every 10 epochs to checkpoints/
#       so ckpt_epoch_010.pth, ckpt_epoch_020.pth, ckpt_epoch_030.pth
#       will be saved to your Drive automatically


[train] Building dataloader...
[dataset] CAMO / Train — 1000 samples found.
[dataset] COD10K / Train — 6000 samples found.
[dataset] Combined dataset — 7000 total samples (1000 + 6000)
/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:424: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  self.check_worker_number_rationality()
[build] Train loader — 7000 samples, 875 batches per epoch (batch size 8)
[train] Building model...

model.safetensors: downloading bytes:  81% 81.7M/101M [00:01<00:00, 109MB/s, 5.19MB/s  ] 
model.safetensors: downloading bytes: 100% 95.6M/95.6M [00:01<00:00, 58.4MB/s, 9.10MB/s  ]
model.safetensors: reconstructing file: 100% 101M/101

DAY1:
What was done
*   Mounted Google Drive and set runtime to T4 GPU
*   Unzipping cod_hybrid.zip directly to Drive failed midway — Drive API rate limits caused silent failure. Uploaded code files manually instead.
*   Fixed DATA_ROOT path in config to match actual Drive folder structure
*   GPU not assigned on first attempt (CPU fallback) — resolved by reconnecting session
*  Training started successfully on T4 GPU with 7000 samples (CAMO + COD10K), loss decreased from 0.172 → 0.067 in ~8 epochs
*  Colab disconnected due to inactivity — no checkpoint saved yet (save interval was every 10 epochs)
*  GPU quota exhausted for the day — resuming tomorrow

Changes for Day 2
*  NUM_WORKERS = 2 — better Drive I/O
*  SAVE_EVERY = 5 — save checkpoint every 5 epochs
*  CAMO only (1000 images) — reduces epoch time from ~27 min to ~4 min, 30 epochs in ~2 hours
*  Run keep-alive cell before training to prevent inactivity disconnection

**DAY2**

In [ ]:
# Cell 1
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Cell 2 - Keep alive (run this before training)
from IPython.display import display, Javascript
display(Javascript('''
function ClickConnect(){
    document.querySelector("colab-toolbar-button#connect").click()
}
setInterval(ClickConnect, 60000)
'''))

<IPython.core.display.Javascript object>

In [ ]:
# Cell 2 - Install timm
!pip install timm -q

In [ ]:
# Cell 3 - Navigate to project
import os
os.chdir('/content/drive/MyDrive/cod_hybrid_project/cod_hybrid')
#!ls

In [ ]:
!python train.py


[train] Building dataloader...
[dataset] CAMO / Train — 1000 samples found.
[build] Train loader — 1000 samples, 125 batches per epoch (batch size 8)
[train] Building model...

model.safetensors: downloading bytes:  79% 80.7M/101M [00:02<00:00, 67.2MB/s, 5.85MB/s  ]
model.safetensors: downloading bytes: 100% 95.6M/95.6M [00:02<00:00, 37.0MB/s, 8.92MB/s  ]
model.safetensors: reconstructing file: 100% 101M/101M [00:02<00:00, 39.2MB/s, 9.59MB/s  ]
[misc] Trainable parameters: 25.65M

[trainer] Starting training for 30 epochs
[trainer] Device: cuda
[trainer] Batch size: 8
[trainer] LR (new modules): 0.0001
[trainer] LR (backbone): 1e-05

  [Epoch 001]  batch 50/125  loss: 0.2279  bce: 0.1678  ual: 0.0601
  [Epoch 001]  batch 100/125  loss: 0.1922  bce: 0.1411  ual: 0.0511
[Epoch 001/30]  loss: 0.1832  bce: 0.1343  ual: 0.0489
  [Epoch 002]  batch 50/125  loss: 0.1349  bce: 0.0966  ual: 0.0382
  [Epoch 002]  batch 100/125  loss: 0.1342  bce: 0.0977  ual: 0.0365
[Epoch 002/30]  loss: 0.1328

In [ ]:
# Cell 3 - Run evaluation on CPU
!python eval.py --checkpoint checkpoints/ckpt_epoch_025.pth --device cpu
# Why: GPU quota is exhausted but eval is lightweight enough for CPU
# What: loads the epoch 25 checkpoint and runs inference on CAMO test set reports BCE loss — our Phase 1 evaluation metric
#       runs one image at a time (batch size 1) so CPU handles it fine


[eval] Building model...
[eval] Loading checkpoint: checkpoints/ckpt_epoch_025.pth
[eval] Loaded weights from epoch 25
[dataset] CAMO / Test — 250 samples found.
[build] Eval loader (CAMO) — 250 test samples
/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:1118: UserWarning: 'pin_memory' argument is set as true but no accelerator is found, then device pinned memory won't be used.
  super().__init__(loader)

[evaluator] CAMO results:
  BCE loss : 0.1012
  (Full metrics — MAE, Sm, Em, Fm — active in Phase 2)


[eval] ── Summary ──────────────────────────
  CAMO        BCE: 0.1012
[eval] ────────────────────────────────────



In [ ]:
# Fix eval datasets to CAMO only
'''The issue is evaluating COD10K test set (4000 images) on CPU is too slow — it's trying to evaluate both CAMO and COD10K which will take hours on CPU.'''
config_path = '/content/drive/MyDrive/cod_hybrid_project/cod_hybrid/configs/cod_hybrid.py'

with open(config_path, 'r') as f:
    content = f.read()

content = content.replace(
    "EVAL_DATASETS  = ['CAMO', 'COD10K']",
    "EVAL_DATASETS  = ['CAMO']   # Phase 1: CAMO only — Phase 2 adds COD10K, NC4K, CHAMELEON"
)

with open(config_path, 'w') as f:
    f.write(content)

print('Done')

Done


In [ ]:
# Save eval results to a text file on Drive
results = """
Phase 1 Evaluation Results
===========================
Date        : 27 July 2026
Checkpoint  : ckpt_epoch_025.pth
Dataset     : CAMO test set (250 images)
Device      : CPU

Training Summary:
  Epochs completed : 25 / 30
  Initial loss     : 0.1723
  Final loss       : 0.0399
  Loss reduction   : 76%

Evaluation Results:
  BCE Loss : 0.1012

Notes:
  - Trained on CAMO only (1000 images) due to Colab free tier GPU limits
  - Full metrics (MAE, Sm, Em, Fm) active in Phase 2
  - Phase 2 will train on CAMO + COD10K for 100 epochs
"""

output_path = '/content/drive/MyDrive/cod_hybrid_project/phase1_eval_results.txt'
with open(output_path, 'w') as f:
    f.write(results)

print(f'Results saved to {output_path}')

Results saved to /content/drive/MyDrive/cod_hybrid_project/phase1_eval_results.txt


Day2
*   GPU quota reset — connected to T4 successfully
*   Added keep-alive JavaScript cell to prevent inactivity disconnection
*   Training started with CAMO only (1000 images, 125 batches/epoch, ~4 min/epoch)
*   Loss decreased from 0.172 → 0.040 over 25 epochs (76% reduction)
*   Runtime stopped at epoch 27 — last checkpoint saved at epoch 25
*   GPU quota exhausted again — 25 epochs deemed sufficient for Phase 1
*   First eval attempt failed — COD10K (4000 images) too slow on CPU
*   Fixed config: EVAL_DATASETS = ['CAMO'] — 250 test images only
*   Evaluation completed successfully — BCE Loss: 0.1012 on CAMO test set
*   Results saved to phase1_eval_results.txt on Drive

Phase 1 Status: COMPLETE ✅
*   Training  : 25 epochs, loss 0.172 → 0.040
*   Evaluation: BCE 0.1012 on CAMO test set (250 images)
*   Checkpoint: ckpt_epoch_025.pth saved to Drive

Next: Phase 2 — full RGPU, FPQ, UAL schedule,
*    100 epochs on CAMO + COD10K, full metrics

